hs_composite_resource/models.py

```diff
@@ -65,17 +66,76 @@ def can_be_published(self):
             # url file cannot be published
             if 'url' in lf.extra_data:
                 return False
+            # check for model instance linked to model program
+            if lf.is_model_instance:
+                mi_aggr = lf
+                mp_aggr = mi_aggr.metadata.executed_by
+                if mp_aggr is not None:
+                    if mp_aggr.resource.short_id != mi_aggr.resource.short_id:
+                        # model instance aggregation and model program aggregations are not in the same resource
+                        # the other resource containing the linked model program must have been published
+                        # before this resource can be published
+                        if not mp_aggr.resource.raccess.published:
+                            return False
+                    else:
+                        # model program aggregation and the linked model instance aggregations are in the same resource
+                        continue
+                model_program_url = mi_aggr.metadata.executed_by_url
+                if model_program_url:
+                    # external linked model program url must be a DOI url for this resource to be published
+                    if not model_program_url.startswith("https://doi.org") and \
+                            not model_program_url.startswith("http://doi.org"):
+                        return False
+
         return True
 
-    def set_default_logical_file(self):
-        """sets an instance of GenericLogicalFile to any resource file objects of this instance
-        of the resource that is not already associated with a logical file. """
+    def remove_aggregation_from_file(self, moved_res_file, src_folder, tgt_folder):
+        """removes association with aggregation (fileset or model program) from this
+        moved resource file
+        :param  moved_res_file: an instance of a ResourceFile which has been moved to a different folder
+        :param  src_folder: folder from which the file got moved from
+        :param  tgt_folder: folder to which the file got moved into
+        """
+        if moved_res_file.has_logical_file and (moved_res_file.logical_file.is_fileset or
```

Comment:
I have concerns with putting logical file code in the resource class.  The composite resource already has too much to manage.  I'm also betting the model program/instance logical files share a lot of the same code as a fileset and possibly could extend the fileset to reduce code duplication.